cpp/src/parquet/variant.cc

```diff
@@ -0,0 +1,743 @@
+      return "BooleanTrue";
+    case VariantPrimitiveType::BooleanFalse:
+      return "BooleanFalse";
+    case VariantPrimitiveType::Int8:
+      return "Int8";
+    case VariantPrimitiveType::Int16:
+      return "Int16";
+    case VariantPrimitiveType::Int32:
+      return "Int32";
+    case VariantPrimitiveType::Int64:
+      return "Int64";
+    case VariantPrimitiveType::Double:
+      return "Double";
+    case VariantPrimitiveType::Decimal4:
+      return "Decimal4";
+    case VariantPrimitiveType::Decimal8:
+      return "Decimal8";
+    case VariantPrimitiveType::Decimal16:
+      return "Decimal16";
+    case VariantPrimitiveType::Date:
+      return "Date";
+    case VariantPrimitiveType::Timestamp:
+      return "Timestamp";
+    case VariantPrimitiveType::TimestampNTZ:
+      return "TimestampNTZ";
+    case VariantPrimitiveType::Float:
+      return "Float";
+    case VariantPrimitiveType::Binary:
+      return "Binary";
+    case VariantPrimitiveType::String:
+      return "String";
+    case VariantPrimitiveType::TimeNTZ:
+      return "TimeNTZ";
+    case VariantPrimitiveType::TimestampTZ:
+      return "TimestampTZ";
+    case VariantPrimitiveType::TimestampNTZNanos:
+      return "TimestampNTZNanos";
+    case VariantPrimitiveType::Uuid:
+      return "Uuid";
+    default:
+      return "Unknown";
+  }
+}
+
+std::string variantTypeToString(VariantType type) {
+  switch (type) {
+    case VariantType::OBJECT:
+      return "OBJECT";
+    case VariantType::ARRAY:
+      return "ARRAY";
+    case VariantType::VARIANT_NULL:
+      return "NULL";
+    case VariantType::BOOLEAN:
+      return "BOOLEAN";
+    case VariantType::BYTE:
+      return "BYTE";
+    case VariantType::SHORT:
+      return "SHORT";
+    case VariantType::INT:
+      return "INT";
+    case VariantType::LONG:
+      return "LONG";
+    case VariantType::STRING:
+      return "STRING";
+    case VariantType::DOUBLE:
+      return "DOUBLE";
+    case VariantType::DECIMAL4:
+      return "DECIMAL4";
+    case VariantType::DECIMAL8:
+      return "DECIMAL8";
+    case VariantType::DECIMAL16:
+      return "DECIMAL16";
+    case VariantType::DATE:
+      return "DATE";
+    case VariantType::TIMESTAMP_TZ:
+      return "TIMESTAMP_TZ";
+    case VariantType::TIMESTAMP_NTZ:
+      return "TIMESTAMP_NTZ";
+    case VariantType::FLOAT:
+      return "FLOAT";
+    case VariantType::BINARY:
+      return "BINARY";
+    case VariantType::TIME:
+      return "TIME";
+    case VariantType::TIMESTAMP_NANOS_TZ:
+      return "TIMESTAMP_NANOS_TZ";
+    case VariantType::TIMESTAMP_NANOS_NTZ:
+      return "TIMESTAMP_NANOS_NTZ";
+    case VariantType::UUID:
+      return "UUID";
+    default:
+      return "UNKNOWN";
+  }
+}
+
+VariantMetadata::VariantMetadata(std::string_view metadata) : metadata_(metadata) {
+  if (metadata.size() < 2) {
+    throw ParquetException("Invalid Variant metadata: too short: " +
+                           std::to_string(metadata.size()));
+  }
```

Comment:
I think there are some conditions that should be checked first when creating metadata?
1. byte order
2. whether the header version is 1

Reply:
About (1) I assume it's lsb format here. (2) is a good point